<a href="https://colab.research.google.com/github/pdrzxzz/failing-equipment/blob/main/failing_equipment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.cluster import KMeans

"""   
Objective: Identify which equipment is likely malfunctioning by analyzing the spectral band measurements across 4 equipments.
Each row represents a spectral band; each column represents one equipment.   
"""   
   

In [12]:
df = pd.read_csv('data.csv')
df = df.drop('Unnamed: 0', axis=1)
df = df.rename(columns= lambda x: x.replace('Object ', 'Equipment-'))
df

Equipment-1  Equipment-2  Equipment-3  Equipment-4
0           375          135          458          475
1            57           47           53           73
2           245          267          242          227
3          1472         1494         1462         1582
4           105           66          103          103
5            54           41           62           64
6           193          209          184          235
7           147           93          122          160
8          1102          674          957         1137
9           720         1033          566          874
10          253          143          171          265
11          685          586          750          803
12          488          355          418          570
13          198          187          220          203
14          360          334          337          365
15         1374         1506         1572         1256
16          156          139          147          175

Considering that each row represents a spectral band and each column a equipment:

We need to look at the column with the most distinct values compared to the same row of other columns.

We can easily put that in numbers by calculating the mean and the standard deviation of each column.

In [19]:
df.mean(axis=0)

Equipment-1    469.647059
Equipment-2    429.941176
Equipment-3    460.235294
Equipment-4    503.941176
dtype: float64

In [18]:
print(f'Overall average: {df.mean(axis=0).mean()}')

Overall average: 465.94117647058823


It seems that Equipment 2 and Equipment 4 have significant differences compared to the absolute mean (i.e., the mean of the column means, which is 465.9):

Equipment 2 has a mean value of 429.9, resulting in a difference of
465.9 − 429.9 = 36.0

Equipment 4 has a mean value of 503.9, resulting in a difference of
503.9 − 465.9 = 38.0

Both deviations are notable when compared to the overall average.

In [11]:
df.std(axis=0)

Equipment-1    452.701466
Equipment-2    479.896014
Equipment-3    469.595907
Equipment-4    463.859282
dtype: float64

Equipment-2 has the highest standard deviation among all, which indicates that its 17 band values vary more compared to the others.    

Final decision: **Equipment-2** is likely to be the failing one
due to its significantly higher variation across spectral bands (highest standard deviation),
and also because its average measurement is notably lower than the overall mean.